Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h

```diff
@@ -0,0 +1,64 @@
+ *
+ */
+
+#pragma once
+
+#include <FeatureMatrix.h>
+#include <FeatureMatrixTransformer.h>
+
+namespace EMotionFX::MotionMatching
+{
+    //! The standard scaler can be used to normalize the feature matrix, the query vector and other data.
+    //! It standardizes features by removing the mean and scaling to the unit variance.
+    //! This implementation is mimicing the behavior of the standard scaler from scikit-learn (sklearn.preprocessing.StandardScaler).
+    //! As we use floats by default, our errors are a bigger, especially if the variance is small as this leads to a division by a small
+    //! value. In case the calculated standard deviation for a given feature is smaller than the given s_epsilon value, the standard
+    //! deviation gets force-set to 1.0 to avoid divisions by infinity and preserve the value when doing the transform -> inverse-transform
+    //! roundtrip.
+    class StandardScaler : public FeatureMatrixTransformer
+    {
+    public:
+        AZ_RTTI(StandardScaler, "{A0C7F056-94B0-43A1-8D12-B1A7B67AB92A}", FeatureMatrixTransformer);
+        AZ_CLASS_ALLOCATOR_DECL;
+
+        StandardScaler() = default;
+
+        bool Fit(const FeatureMatrix& featureMatrix, const Settings& settings = {}) override;
+
+        // Normalize and scale the values.
+        float Transform(float value, FeatureMatrix::Index column) const override;
+        AZ::Vector2 Transform(const AZ::Vector2& value, FeatureMatrix::Index column) const override;
+        AZ::Vector3 Transform(const AZ::Vector3& value, FeatureMatrix::Index column) const override;
+        void Transform(AZStd::span<float> data) const override;
+        FeatureMatrix Transform(const FeatureMatrix& featureMatrix) const override;
+
+        // From normalized and scaled back to the original values.
+        FeatureMatrix InverseTransform(const FeatureMatrix& featureMatrix) const override;
+        AZ::Vector2 InverseTransform(const AZ::Vector2& value, FeatureMatrix::Index column) const override;
+        AZ::Vector3 InverseTransform(const AZ::Vector3& value, FeatureMatrix::Index column) const override;
+        float InverseTransform(float value, FeatureMatrix::Index column) const override;
+
+        const AZStd::vector<float>& GetMeans() const
+        {
+            return m_means;
+        }
+        const AZStd::vector<float>& GetStandardDeviations() const
+        {
+            return m_standardDeviations;
+        }
+
+        static constexpr float s_epsilon = AZ::Constants::FloatEpsilon;
+
+        void SaveAsCsv(const char* filename, const AZStd::vector<AZStd::string>& columnNames = {});
+
+    private:
+        AZStd::vector<float> m_means; //!< The mean value for each feature / column.
+        AZStd::vector<float> m_standardDeviations; //!< The standard deviation for each feature / column.
+        static constexpr float s_deltaDegreesOfFreedom = 0.0f; // Used to set the bias of the estimator. [Default=unbiased]
```

Comment:
I think there is a bit of subtlety here depending on whether you're talking about the sample standard deviation or the estimate of the population standard deviation. When you talk about bias, it's about whether the expected error between your estimate of the population standard deviation and the ground truth is zero (unbiased) or non-zero (biased). If you set the DOF to 0, you actually will have a biased estimate of the population standard deviation (it will on average be slightly too small because on average the sample values will be slightly closer to the sample mean than they are to the population mean). To get an unbiased estimate of the population standard deviation, you actually would want DOF = 1. 

Having said all that:
(a) I think it's perfectly reasonable to scale based on the sample mean and standard deviation
(b) if the sample size is even remotely large, the difference is basically going to be negligible

To summarize, I think strictly speaking referring to DOF = 0 as unbiased is not correct. But since I think the effect is going to be insignificant, I wonder if you really need the DOF term at all or if it could just be completely omitted?

Reply:
Removed as it is currently unused.